File: vol2/ch20/LeetCode1102-2.cpp
struct QNode {                  //优先队列结点类型
    int x,y;                    //位置
    int length;                 //路径长度
    bool operator<(const QNode& b) const {
        return length<b.length;     //length越大越优先出队
    }
};
class Solution {
    int dx[4]={0,0,1,-1};       //水平方向偏移量
    int dy[4]={1,-1,0,0};       //垂直方向偏移量
public:
    int maximumMinimumPath(vector<vector<int>>& grid) {
        int m=grid.size(),n=grid[0].size();
        vector<vector<int>> dist(m,vector<int>(n,-1));
        priority_queue<QNode> qu;
        QNode e,e1;
        e.x=0;e.y=0;e.length=grid[0][0];
        qu.push(e);
        dist[0][0]=grid[0][0];
        int ans=0;
        while(!qu.empty()) {
            e=qu.top();qu.pop();      //出队元素e
            int x=e.x,y=e.y,length=e.length;
            if(x==m-1 && y==n-1)        //找到目标
                ans=max(ans,length);
            for(int di=0;di<4;di++) {
                int nx=x+dx[di];
                int ny=y+dy[di];
                if(nx<0 || nx>=m || ny<0 || ny>=n) continue;
                int curlength=min(dist[x][y],grid[nx][ny]);
                if(curlength>dist[nx][ny]) {    //剪支
                    dist[nx][ny]=curlength;
                    e1.x=nx;e1.y=ny;e1.length=dist[nx][ny];
                    qu.push(e1);
                }
            }
        }
        return ans;
    }
};

File: vol2/ch20/LeetCode1102-2.py
class Solution:
    def maximumMinimumPath(self, grid: List[List[int]]) -> int:
        dx=[0,0,1,-1]                       #水平方向偏移量
        dy=[1,-1,0,0]                   #垂直方向偏移量
        m,n=len(grid),len(grid[0])
        dist=[[-1 for _ in range(0,n)] for _ in range(0,m)]
        pqu=[]                          #定义大根堆pqu,元素为[-length,x,y]
        heapq.heappush(pqu,[-grid[0][0],0,0])      #源点元素进队
        dist[0][0]=grid[0][0]
        ans=0
        while pqu:
            e=heapq.heappop(pqu)        #出队元素e
            length,x,y=e[0],e[1],e[2]
            length=-length              #恢复正值
            if x==m-1 and y==n-1:ans=max(ans,length)
            for di in range(0,4):
                nx,ny=x+dx[di],y+dy[di]
                if nx<0 or nx>=m or ny<0 or ny>=n:continue
                curlength=min(dist[x][y],grid[nx][ny])
                if curlength>dist[nx][ny]:
                    dist[nx][ny]=curlength
                    heapq.heappush(pqu,[-dist[nx][ny],nx,ny])
        return ans
